### PR TITLE
Restore window state when diagram is loaded

### PR DESCRIPTION
**Pull request by amolenaar**
### PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes

### What is the current behavior?

All models (windows) use one central properties configuration. Who last writes, wins.

Issue Number: #710 

### What is the new behavior?

Settings are saved per model. Settings for the no-model state (empty file name) are also saved, so when you start Gaphor again, it starts in the state it's left in.

* Added state for open/close of element editor and opened diagrams.
* services are not maintained in a list in ComponentRegistry, so they can be shut down in a predictable order. This help with storing the right opened diagrams.
* Session creation logic has moved from AppFileManager to Application and Session.
* File loading is now asynchronous with use of `g_async` (from Gaphas)
* Window and component size is not stored when the window is maximized or in full screen
* After initialization a `SessionCreated` event is sent, both in the session and on application level.
* Updated Gaphas to 3.1.4

### Does this PR introduce a breaking change?
- [X] Yes
- [ ] No

### CODE REVIEW

**Review by danyeaw — APPROVED** [review]
This looks really good! This is a really nice feature to load things back the way they were.

I noticed this change got rid of the file is loading status bar, was that on purpose?
Also this might be for a future enhancement, but do we want to save where on the screen the window was when it was closed?

**Comment by amolenaar** [comment]
> This looks really good! This is a really nice feature to load things back the way they were.
> 
> I noticed this change got rid of the file is loading status bar, was that on purpose?

The loading bar is not shown when the initial model is loaded. The initial "open" action (`gaphor/ui/__init__.py`) is not executed from within the GLib main loop. When you open a model from the "Open" file menu, a status bar is shown, because then it is loaded from within the main loop.

`g_async` does not schedule an idle task if it's not in a main loop, on purpose ;).

I figured I leave it like this for now.

> Also this might be for a future enhancement, but do we want to save where on the screen the window was when it was closed?

I think at some point we want that (the coordinates, zoom, etc.). Also we probably want the restore active diagram. Currently the last diagram opened will be focused.

File: gaphor/services/properties.py [hunk 1]
```diff
@@ -28,6 +30,22 @@ def get_config_dir() -> str:
...
+    """
+
+    cache_dir = os.path.join(GLib.get_user_cache_dir(), "gaphor")
+    os.makedirs(cache_dir, exist_ok=True)
+
+    return cache_dir
+
+
+def file_hash(filename: str) -> str:
+    return hashlib.blake2b(filename.encode("utf-8"), digest_size=24).hexdigest()
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/services/properties.py:46]: Me neither, until I looked into `hashlib` 😃 .

File: gaphor/services/properties.py [hunk 2]
```diff
@@ -28,6 +30,22 @@ def get_config_dir() -> str:
     return config_dir
 
 
+def get_cache_dir() -> str:
+    """Return the directory where the user's cache is stored.
+
+    This varies depending on platform.
+    """
+
+    cache_dir = os.path.join(GLib.get_user_cache_dir(), "gaphor")
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/services/properties.py:39]: I stuck with the way the config-dir was working. Pathlib is nicer to use, probably.

Maybe a future improvement?

**Comment by danyeaw** [comment]
> The loading bar is not shown on startup when a model is loaded: the initial "open" action (`gaphor/ui/__init__.py`) is not executed from within the GLib main loop. When you open a model from the "Open" file menu, a status bar is shown, because then it is loaded from within the main loop.

Ahh, I see it now. I think the issue is that it is showing up behind the main Window for me.

image

File: gaphor/services/properties.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (danyeaw) [reply gaphor/services/properties.py:39]: Yup, sounds good!

**Comment by amolenaar** [comment]
> Ahh, I see it now. I think the issue is that it is showing up behind the main Window for me.

Fixed. Now the progress bar should stay on top.

**Comment by danyeaw** [comment]
> Fixed. Now the progress bar should stay on top.

Yup, works great now :+1:

**Comment by amolenaar** [comment]
Can it be merged then?

**Pull request merged**